content/blog/2019-09-25-bril2jb.md

```diff
@@ -0,0 +1,75 @@
++++
+title = "bril2jb: A tool that translates bril to Java Bytecode"
+extra.author = "Siqiu Yao"
+extra.bio = """
+  [Siqiu Yao](http://www.cs.cornell.edu/~yaosiqiu/) is a 3rd year Ph.D. student interested in security about systems and programming languages.~~~~~~.
+"""
++++
+
+### Goal
+The goal of this project is to provide a tool to translate bril code to Java Bytecode.
+
+### Design
+#### Before the actual translation process
+At first glance, translating bril to [Java Bytecode](https://en.wikipedia.org/wiki/Java_bytecode) seems a 
+pretty straightforward job, since bril (currently) only contains simple instructions and supports no function invocation.
+But it turns out that making something directly runnable by *Java Virtual Machine*(JVM, the run-time environment of Java Bytecode)
+is not trivial, because we need to construct our program as a Java *class*, 
+which is the only code format JVM accepts.
+
+Therefore we need to create a wrapper class(say `Wrapper.class`) like the following :
+```JAVA
+public class Wrapper {
+  public Wrapper();
+  public static void main(java.lang.String[]);
+}
+```
+And the `main` function would contain the instructions in the bril code.
+
+#### The translation process
+I chose [ASM](https://asm.ow2.io), a Java Bytecode manipulation and analysis framework to help with the translation.
+
+The translation process is pretty regular. One interesting point is that Java Bytecode is strictly-typed, 
+while bril is naturally dynamic and 
+there might be no explicit declaration of a variable before it is used(in the order of instruction list),
+and some instructions are polymorphic such as `print`.
+Therefore, we preprocess the instructions first to gather all type information and labels.
+This step also infers types of variables, so there is actually no need to indicate types in bril anymore.
```

Comment:
I’m not quite sure what you mean by this. What do you mean by saying you “inferred” the types? Did you just look for the assignments to the variables and use their types as written in the instructions that assign to them? (That’s not really inference; it’s just writing down the types.) Or did you need something more global?

Please expand this discussion with more detail.